### PR TITLE
Derivation of subset SPLs. Bachelor's thesis implementation.

### PR DESCRIPTION
**Pull request by PaulWestphal**

### CODE REVIEW

**Comment by PaulWestphal** [comment]
todo:

- fix capitalization in context menu entry
- give unique name (timestamp) to subset SPL to avoid errors 
- do not add root to new constraint with all selections
- new constraint should be at the bottom of all constraints

**Comment by tthuem** [comment]
Please fix capitalization also for "Set as current configuration". Is it feasible to ignore core features when adding the constraint?

**Comment by PaulWestphal** [comment]
Good idea. Yes, it's feasible and I implemented it.

File: plugins/de.ovgu.featureide.core.featurehouse/src/de/ovgu/featureide/featurehouse/FeatureHouseComposer.java [hunk 1]
```diff
@@ -1210,4 +1211,41 @@ public boolean needColor() {
...
+		return true;
+	}
+
+	@Override
+	public void buildPartialFeatureProjectAssets(IFolder sourceFolder, ArrayList<String> removedFeatures, ArrayList<String> mandatoryFeatures)
+			throws IOException, CoreException {
+		deleteFeatureFolders(sourceFolder, removedFeatures);
+	}
+
+	private void deleteFeatureFolders(IFolder sourceFolder, ArrayList<String> removedFeatures) {
```
Comment (SundermannC) [inline plugins/de.ovgu.featureide.core.featurehouse/src/de/ovgu/featureide/featurehouse/FeatureHouseComposer.java:1225]: Seems to complicated. Is it necessary to create the intermediate lists?

File: plugins/de.ovgu.featureide.core.antenna/src/de/ovgu/featureide/antenna/AntennaPreprocessor.java [hunk 2]
```diff
@@ -698,4 +697,359 @@ private boolean isAnnotation(String line) {
...
+					break;
+				}
+			}
+
+			if (!containsDeletedFeature) {
+				// node stayed the same
+				annotationDecision.add(ANNOTATION_KEPT);
+			} else if (!((afterNode instanceof True) || (afterNode instanceof False))) {
+				// has a solution and is not a tautology
+				// annotation anpassen, dafür afternode zu string verarbeiten
```
Comment (SundermannC) [inline plugins/de.ovgu.featureide.core.antenna/src/de/ovgu/featureide/antenna/AntennaPreprocessor.java:795]: Seems like a TODO

File: plugins/de.ovgu.featureide.core.antenna/src/de/ovgu/featureide/antenna/AntennaPreprocessor.java [hunk 3]
```diff
@@ -698,4 +697,359 @@ private boolean isAnnotation(String line) {
...
+					}
+				} else if (block instanceof IfBlock) {
+					lines.set(block.getStartLine(), translateNodeToAntennaStatement(afterNode, false));
+					annotationDecision.add(ANNOTATION_KEPT);
+				} else if (block instanceof ElseBlock) {
+					annotationDecision.add(ANNOTATION_KEPT);
+				}
+			} else if (new SatSolver(beforeNode, 1000).hasSolution() && (afterNode instanceof False)) {
+				// removing features causes this annotation to be a contradiction
+				// gesamten codeblock entfernen
```
Comment (SundermannC) [inline plugins/de.ovgu.featureide.core.antenna/src/de/ovgu/featureide/antenna/AntennaPreprocessor.java:845]: Seems like a TODO

File: plugins/de.ovgu.featureide.core.antenna/src/de/ovgu/featureide/antenna/AntennaPreprocessor.java [hunk 4]
```diff
@@ -698,4 +697,359 @@ private boolean isAnnotation(String line) {
...
+				for (int line = block.getStartLine(); line <= (block.getEndLine() - 1); line++) {
+					lines.set(line, "");
+				}
+				annotationDecision.add(ANNOTATION_AND_BLOCK_REMOVED);
+			} else if (new SatSolver(new Not(beforeNode), 1000).hasSolution() && (afterNode instanceof True)) {
+				// removing features causes this annotation to be a tautology
+				// annotation entfernen
```
Comment (SundermannC) [inline plugins/de.ovgu.featureide.core.antenna/src/de/ovgu/featureide/antenna/AntennaPreprocessor.java:852]: Seems like a TODO

File: plugins/de.ovgu.featureide.core.antenna/src/de/ovgu/featureide/antenna/AntennaPreprocessor.java [hunk 5]
```diff
@@ -698,4 +697,359 @@ private boolean isAnnotation(String line) {
...
+		for (final int decision : annotationDecision) {
+			if (decision != ANNOTATION_KEPT) {
+				return true;
+			}
+		}
+
+		return changed;
+	}
+
+	/**
```
Comment (SundermannC) [inline plugins/de.ovgu.featureide.core.antenna/src/de/ovgu/featureide/antenna/AntennaPreprocessor.java:923]: Outdated Java docs for multiple methods

**Review by SundermannC — COMMENTED** [review]
Minor changes requested. Otherwise, the code seems fine.

File: plugins/de.ovgu.featureide.core.featurehouse/src/de/ovgu/featureide/featurehouse/FeatureHouseComposer.java (hunk 1 above)
Comment (PaulWestphal) [reply plugins/de.ovgu.featureide.core.featurehouse/src/de/ovgu/featureide/featurehouse/FeatureHouseComposer.java:1225]: You're right, it's unnecessary.

File: plugins/de.ovgu.featureide.core.antenna/src/de/ovgu/featureide/antenna/AntennaPreprocessor.java (hunk 2 above)
Comment (PaulWestphal) [reply plugins/de.ovgu.featureide.core.antenna/src/de/ovgu/featureide/antenna/AntennaPreprocessor.java:795]: Oops.

**Pull request merged**